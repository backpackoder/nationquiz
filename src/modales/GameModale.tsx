import { useNavigate } from "react-router-dom";
import { useTranslation } from "react-i18next";

type GameModaleProps = {
  setIsModaleOpened: React.Dispatch<React.SetStateAction<boolean>>;
  gameModale: any;
  gameDispatch: any;
};

export function GameModale({ setIsModaleOpened, gameModale, gameDispatch }: GameModaleProps) {
  const navigate = useNavigate();
  const { t } = useTranslation();
  const { confirmation, description } = gameModale;

  return (
    <>
      <p className="description">{description}</p>

      <div className="buttons">
        <button
          onClick={() =>
            confirmation === "restart" ? gameDispatch({ type: confirmation }) : navigate("/quiz")
          }
        >
          {t("modale.game.accept")}
        </button>

        <button onClick={() => setIsModaleOpened(false)}>{t("modale.game.decline")}</button>
      </div>
    </>
  );
}
