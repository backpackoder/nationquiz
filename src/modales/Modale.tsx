// Icons
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faXmark } from "@fortawesome/free-solid-svg-icons";

// Types
import { ModaleProps } from "../types/props";

export function Modale({ name, children, setIsModaleOpened }: ModaleProps) {
  function handleModalClick(e: React.MouseEvent<HTMLDivElement>) {
    e.stopPropagation();
  }

  return (
    <div className={"modaleWrapper"} onClick={() => setIsModaleOpened(false)}>
      <div className={`modale ${name}`} onClick={(e) => handleModalClick(e)}>
        <FontAwesomeIcon
          icon={faXmark}
          size="2xl"
          className="closeBtn"
          onClick={() => setIsModaleOpened(false)}
        />
        {children}
      </div>
    </div>
  );
}
