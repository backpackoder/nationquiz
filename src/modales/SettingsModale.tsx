import { useContext, useEffect } from "react";
import { AppContext } from "../AppContext";
import { useLocation, useNavigate } from "react-router-dom";
import { useTranslation } from "react-i18next";

// Types
import { SettingsModaleProps } from "../types/props";
import { Regions, SettingEnum } from "../types/settings";

// Hooks
import { AppProviderProps } from "../types/context";

// Commons
import { CONTINENTS, ROUTES, THEMES } from "../commons/commons";

export function SettingsModale({ quizTheme, setIsModaleOpened }: SettingsModaleProps) {
  const navigate = useNavigate();
  const { settingsDispatch, settingsList }: AppProviderProps = useContext(AppContext);
  const { t } = useTranslation();
  const { pathname } = useLocation();

  const difficulty = settingsList[SettingEnum.Difficulty];
  const quizLength = settingsList[SettingEnum.QuizLength];
  const region = settingsList[SettingEnum.Region];

  const T_ROOT = "modale.settings";

  function goToQuiz() {
    setIsModaleOpened(false);
    navigate(
      pathname === ROUTES.HOME ? `${ROUTES.QUIZ.ROOT}${quizTheme.theme}` : `${quizTheme.theme}`
    );
  }

  useEffect(() => {
    switch (quizTheme.theme) {
      case THEMES.BORDERS:
        if (region.value === CONTINENTS.OCEANIA) {
          settingsDispatch({
            type: "region not available for this theme",
            payload: { value: CONTINENTS.WORLD },
          });
        }
        break;

      default:
        break;
    }
  }, [quizTheme]);

  return (
    <>
      <h2>{t(`quizList.${quizTheme.theme}.title`)}</h2>

      <h3>{t(`${T_ROOT}.title`)}</h3>

      <div className="settings">
        <div className={`setting ${difficulty.setting.title}`}>
          <h4>{t(`${T_ROOT}.${difficulty.setting.title}.title`)}</h4>

          <div className="buttons">
            {difficulty.setting.values.map((item, index) => {
              return (
                <button
                  key={index}
                  className={difficulty.value === item.value ? "active" : "inactive"}
                  onClick={() =>
                    settingsDispatch({
                      type: difficulty.callDispatch,
                      payload: { value: item.value },
                    })
                  }
                >
                  {item.label}
                </button>
              );
            })}
          </div>
        </div>

        <div className={`setting ${quizLength.setting.title}`}>
          <h4>{t(`${T_ROOT}.${quizLength.setting.title}.title`)}</h4>

          <div className="buttons">
            {quizLength.setting.values.map((item, index) => {
              return (
                <button
                  key={index}
                  className={quizLength.value === item.value ? "active" : "inactive"}
                  onClick={() =>
                    settingsDispatch({
                      type: quizLength.callDispatch,
                      payload: { value: item.value },
                    })
                  }
                >
                  {item.label}
                </button>
              );
            })}
          </div>
        </div>

        <div className={`setting ${region.setting.title}`}>
          <h4>{t(`${T_ROOT}.${region.setting.title}.title`)}</h4>

          <div className="buttons">
            {region.setting.values.map((item, index) => {
              return (
                <button
                  key={index}
                  className={
                    quizTheme.theme === THEMES.BORDERS && index === Regions.Oceania
                      ? "hidden"
                      : region.value === item.value
                      ? "active"
                      : "inactive"
                  }
                  onClick={() =>
                    settingsDispatch({
                      type: region.callDispatch,
                      payload: { value: item.value },
                    })
                  }
                >
                  {item.label}
                </button>
              );
            })}
          </div>
        </div>
      </div>

      <button className="start" onClick={() => goToQuiz()}>
        {t(`${T_ROOT}.start`)}
      </button>
    </>
  );
}
