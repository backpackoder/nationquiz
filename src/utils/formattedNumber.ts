type formattedNumber = {
  number: number;
  language: string;
};

export function getFormattedNumber({ number, language }: formattedNumber) {
  const billions = Math.floor(number / 1_000_000_000);
  const millions = Math.floor((number - billions * 1_000_000_000) / 1_000_000);
  const thousands = Math.floor((number - billions * 1_000_000_000 - millions * 1_000_000) / 1000);
  const ones = number - billions * 1_000_000_000 - millions * 1_000_000 - thousands * 1000;

  const formattedBillions = billions.toLocaleString(language);
  const formattedMillions = millions.toLocaleString(language);
  const formattedThousands = thousands.toLocaleString(language);
  const formattedOnes = ones.toLocaleString(language);

  const parts = [];
  if (billions > 0) {
    parts.push(formattedBillions);
  }
  if (millions > 0) {
    parts.push(formattedMillions.padStart(parts.length > 0 ? 3 : 0, "0"));
  }
  if (thousands > 0) {
    parts.push(formattedThousands.padStart(parts.length > 0 ? 3 : 0, "0"));
  }
  if (ones >= 0) {
    parts.push(formattedOnes.padStart(parts.length > 0 ? 3 : 0, "0"));
  }

  const result = parts.join(language === "en" ? "." : ",");

  return result;
}
