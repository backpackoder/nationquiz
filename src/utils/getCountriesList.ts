// Types
import { Regions } from "../types/settings";

// Commons
import { CONTINENTS, THEMES } from "../commons/commons";
import { API_DATA } from "../types/api";

export type GetCountriesList = {
  data: API_DATA[];
  theme: string | undefined;
  region: string;
};

export function getCountriesList({ data, theme, region }: GetCountriesList) {
  function getTheme(): API_DATA[] {
    switch (theme) {
      case THEMES.FLAGS:
        return data.filter((country: API_DATA) => country.cca2 !== undefined);

      case THEMES.CAPITALS:
        return data.filter((country: API_DATA) => country.capital !== undefined);

      case THEMES.DEMOGRAPHY:
        return data.filter((country: API_DATA) => country.population !== undefined);

      case THEMES.BORDERS:
        return data.filter((country: API_DATA) => country.borders !== undefined);

      case THEMES.AREAS:
        return data.filter((country: API_DATA) => country.area !== undefined);

      default:
        throw new Error("Theme not found");
    }
  }

  switch (region) {
    case CONTINENTS.AFRICA:
      return getTheme().filter((country: API_DATA) => country.region === CONTINENTS.AFRICA);

    case CONTINENTS.AMERICAS:
      return getTheme().filter((country: API_DATA) => country.region === CONTINENTS.AMERICAS);

    case CONTINENTS.ASIA:
      return getTheme().filter((country: API_DATA) => country.region === CONTINENTS.ASIA);

    case CONTINENTS.EUROPE:
      return getTheme().filter((country: API_DATA) => country.region === CONTINENTS.EUROPE);

    case CONTINENTS.OCEANIA:
      return getTheme().filter((country: API_DATA) => country.region === CONTINENTS.OCEANIA);

    default:
      return getTheme();
  }
}
