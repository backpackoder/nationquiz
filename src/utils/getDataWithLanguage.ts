import { API_DATA } from "../types/api";

export type GetDataDependingOnLanguage = {
  data: API_DATA;
  type: GetDataDependingOnLanguageType;
  language: string;
};

export type GetDataDependingOnLanguageType = "common name" | "official name";

export function getDataDependingOnLanguage({ data, type, language }: GetDataDependingOnLanguage) {
  const languageIsntEnglish = language === "fra" || language === "spa";

  switch (type) {
    case "common name":
      return languageIsntEnglish ? data.translations[language].common : data.name.common;

    case "official name":
      return languageIsntEnglish ? data.translations[language].official : data.name.official;

    default:
      throw new Error("Invalid data");
  }
}
