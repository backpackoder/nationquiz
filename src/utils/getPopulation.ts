type FormatedPopulation = {
  population: number;
  language: string;
};

export function formattedPopulation({ population, language }: FormatedPopulation) {
  const billions = Math.floor(population / 1_000_000_000);
  const millions = Math.floor((population - billions * 1_000_000_000) / 1_000_000);
  const thousands = Math.floor(
    (population - billions * 1_000_000_000 - millions * 1_000_000) / 1000
  );
  const ones = population - billions * 1_000_000_000 - millions * 1_000_000 - thousands * 1000;

  const formattedBillions = billions.toLocaleString(language);
  const formattedMillions = millions.toLocaleString(language);
  const formattedThousands = thousands.toLocaleString(language);
  const formattedOnes = ones.toLocaleString(language);

  const parts = [];
  if (billions > 0) {
    parts.push(formattedBillions);
  }
  if (millions > 0) {
    parts.push(formattedMillions.padStart(parts.length > 0 ? 3 : 0, "0"));
  }
  if (thousands > 0) {
    parts.push(formattedThousands.padStart(parts.length > 0 ? 3 : 0, "0"));
  }
  if (ones > 0) {
    parts.push(formattedOnes.padStart(parts.length > 0 ? 3 : 0, "0"));
  }

  const result = parts.join(language === "en" ? "." : ",");

  return result;
}
