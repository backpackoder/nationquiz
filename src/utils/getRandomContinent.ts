// Commons
import { CONTINENTS } from "../commons/commons";

export function getRandomContinent() {
  const continents_list = Object.keys(CONTINENTS).map(
    (key) => CONTINENTS[key as keyof typeof CONTINENTS]
  );
  const randomContinent = continents_list[Math.floor(Math.random() * continents_list.length)];

  return randomContinent;
}
