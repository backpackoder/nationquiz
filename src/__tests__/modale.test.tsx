import { render, screen } from "./utils/test-utils";
import { t } from "i18next";

// Components
import { SettingsModale } from "../modales/SettingsModale";

const quizTheme = { theme: "flags", title: "title", description: "description" };
const setIsModaleOpened = () => {};

test("modale", () => {
  render(<SettingsModale quizTheme={quizTheme} setIsModaleOpened={setIsModaleOpened} />);

  const quizThemeTitle = t(`quizList.${"flags"}.title`);
  const btn = t("modale.settings.start");
  const kid = t("modale.settings.difficulty.kid");
  const short = `${t("modale.settings.length.short")}`;
  const world = t("modale.settings.region.world");

  expect(screen.getByText(quizThemeTitle)).toBeInTheDocument();
  expect(screen.getByText(btn)).toBeInTheDocument();
  expect(screen.getByText(kid)).toBeInTheDocument();
  expect(screen.getByText(short)).toBeInTheDocument();
  expect(screen.getByText(world)).toBeInTheDocument();
});
