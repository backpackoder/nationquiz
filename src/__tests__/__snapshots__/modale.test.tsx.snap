// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`modale 1`] = `
<div>
  <h2>
    Flags
  </h2>
  <h3>
    Settings
  </h3>
  <div
    class="settings"
  >
    <div
      class="setting difficulty"
    >
      <h4>
        Difficulty
      </h4>
      <div
        class="buttons"
      >
        <button
          class="inactive"
        >
          Kid
        </button>
        <button
          class="active"
        >
          Easy
        </button>
        <button
          class="inactive"
        >
          Medium
        </button>
        <button
          class="inactive"
        >
          Hard
        </button>
        <button
          class="inactive"
        >
          Expert
        </button>
      </div>
    </div>
    <div
      class="setting length"
    >
      <h4>
        Questions
      </h4>
      <div
        class="buttons"
      >
        <button
          class="inactive"
        >
          short
           (
          short
          )
        </button>
        <button
          class="active"
        >
          normal
           (
          normal
          )
        </button>
        <button
          class="inactive"
        >
          long
           (
          long
          )
        </button>
      </div>
    </div>
    <div
      class="setting region"
    >
      <h4>
        Continents
      </h4>
      <div
        class="buttons"
      >
        <button
          class="active"
        >
          All
        </button>
        <button
          class="inactive"
        >
          Africa
        </button>
        <button
          class="inactive"
        >
          Americas
        </button>
        <button
          class="inactive"
        >
          Asia
        </button>
        <button
          class="inactive"
        >
          Europe
        </button>
        <button
          class="inactive"
        >
          Oceania
        </button>
      </div>
    </div>
  </div>
  <button
    class="start"
  >
    Start the game
  </button>
</div>
`;
